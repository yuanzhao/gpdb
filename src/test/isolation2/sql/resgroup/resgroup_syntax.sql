-- ----------------------------------------------------------------------
-- Test: assign/alter a role to a resource group
-- ----------------------------------------------------------------------

--start_ignore
DROP ROLE IF EXISTS rg_test_role;
--end_ignore
CREATE ROLE rg_test_role;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role RESOURCE GROUP non_exist_group;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role RESOURCE GROUP admin_group;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role RESOURCE GROUP none;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';

DROP ROLE rg_test_role;
CREATE ROLE rg_test_role SUPERUSER;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role RESOURCE GROUP default_group;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role RESOURCE GROUP none;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';

DROP ROLE rg_test_role;
CREATE ROLE rg_test_role RESOURCE GROUP non_exist_group;

-- nonsuper user should not be assigned to admin group
CREATE ROLE rg_test_role RESOURCE GROUP admin_group;


-- ----------------------------------------------------------------------
-- Test: create/drop a resource group
-- ----------------------------------------------------------------------

--start_ignore
DROP RESOURCE GROUP rg_test_group;
--end_ignore

-- can't drop non-exist resource group
DROP RESOURCE GROUP rg_test_group;

-- can't create the reserved resource groups
CREATE RESOURCE GROUP default_group WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.5, memory_redzone_limit=.7);
CREATE RESOURCE GROUP admin_group WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.5, memory_redzone_limit=.7);
CREATE RESOURCE GROUP none WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.5, memory_redzone_limit=.7);

-- must specify both memory_limit and cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, memory_limit=.5, memory_redzone_limit=.7);
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.5, memory_redzone_limit=.7);

CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.6, memory_redzone_limit=.7);
SELECT groupname,concurrency,proposed_concurrency,cpu_rate_limit,memory_limit,proposed_memory_limit,memory_redzone_limit FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';

-- multiple resource groups can't share the same name
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);

-- cpu_rate_limit/memory_limit range is (0.01, 1)
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.5, memory_limit=.05);
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.05, memory_limit=.5);
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.01, memory_limit=.05);
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.05, memory_limit=.01);

-- can't specify the resource limit type multiple times
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, concurrency=1);
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, cpu_rate_limit=.05);
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, memory_limit=.05);
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, memory_redzone_limit=.8);

DROP RESOURCE GROUP rg_test_group;

-- can't drop reserved resource groups
DROP RESOURCE GROUP default_group;
DROP RESOURCE GROUP admin_group;
DROP RESOURCE GROUP none;

-- can't drop non-exist resource group
DROP RESOURCE GROUP rg_non_exist_group;

-- ----------------------------------------------------------------------
-- Test: alter a resource group
-- ----------------------------------------------------------------------
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);

-- ALTER RESOURCE GROUP SET CONCURRENCY N
-- negative
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -2;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -0.5;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0.5;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY a;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 'abc';
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY '1';
-- positive 
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -1;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 1;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 2;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 1000;

DROP RESOURCE GROUP rg_test_group;

-- ----------------------------------------------------------------------
-- Test: create/alter/drop a resource group in subtransaction
-- ----------------------------------------------------------------------

-- CREATE RESOURCE GROUP cannot run inside a subtransaction
BEGIN;
SAVEPOINT rg_savepoint;
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
ROLLBACK TO SAVEPOINT rg_savepoint;
ABORT;

-- ALTER RESOURCE GROUP cannot run inside a subtransaction
BEGIN;
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
SAVEPOINT rg_savepoint;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 10;
ROLLBACK TO SAVEPOINT rg_savepoint;
ABORT;

-- DROP RESOURCE GROUP cannot run inside a subtransaction
BEGIN;
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
SAVEPOINT rg_savepoint;
DROP RESOURCE GROUP rg_test_group;
ROLLBACK TO SAVEPOINT rg_savepoint;
ABORT;
