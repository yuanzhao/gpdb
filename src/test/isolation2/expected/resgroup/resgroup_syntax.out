-- ----------------------------------------------------------------------
-- Test: assign/alter a role to a resource group
-- ----------------------------------------------------------------------

--start_ignore
DROP ROLE IF EXISTS rg_test_role;
DROP
--end_ignore
CREATE ROLE rg_test_role;
CREATE
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
ALTER ROLE rg_test_role RESOURCE GROUP non_exist_group;
ERROR:  resource group "non_exist_group" does not exist
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
ALTER ROLE rg_test_role RESOURCE GROUP admin_group;
ERROR:  only superuser can be assigned to admin resgroup
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
ALTER ROLE rg_test_role RESOURCE GROUP none;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)

DROP ROLE rg_test_role;
DROP
CREATE ROLE rg_test_role SUPERUSER;
CREATE
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6438       
(1 row)
ALTER ROLE rg_test_role RESOURCE GROUP default_group;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
ALTER ROLE rg_test_role RESOURCE GROUP none;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6438       
(1 row)

DROP ROLE rg_test_role;
DROP
CREATE ROLE rg_test_role RESOURCE GROUP non_exist_group;
ERROR:  resource group "non_exist_group" does not exist

-- nonsuper user should not be assigned to admin group
CREATE ROLE rg_test_role RESOURCE GROUP admin_group;
ERROR:  only superuser can be assigned to admin resgroup


-- ----------------------------------------------------------------------
-- Test: create/drop a resource group
-- ----------------------------------------------------------------------

--start_ignore
DROP RESOURCE GROUP rg_test_group;
ERROR:  resource group "rg_test_group" does not exist
--end_ignore

-- can't drop non-exist resource group
DROP RESOURCE GROUP rg_test_group;
ERROR:  resource group "rg_test_group" does not exist

-- can't create the reserved resource groups
CREATE RESOURCE GROUP default_group WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.5, memory_redzone_limit=.7);
ERROR:  resource group "default_group" already exists
CREATE RESOURCE GROUP admin_group WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.5, memory_redzone_limit=.7);
ERROR:  resource group "admin_group" already exists
CREATE RESOURCE GROUP none WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.5, memory_redzone_limit=.7);
ERROR:  resource group name "none" is reserved

-- must specify both memory_limit and cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, memory_limit=.5, memory_redzone_limit=.7);
ERROR:  must specify both memory_limit and cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.5, memory_redzone_limit=.7);
ERROR:  must specify both memory_limit and cpu_rate_limit

CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.5, memory_limit=.6, memory_redzone_limit=.7);
CREATE
SELECT groupname,concurrency,proposed_concurrency,cpu_rate_limit,memory_limit,proposed_memory_limit,memory_redzone_limit FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';
groupname    |concurrency|proposed_concurrency|cpu_rate_limit|memory_limit|proposed_memory_limit|memory_redzone_limit
-------------+-----------+--------------------+--------------+------------+---------------------+--------------------
rg_test_group|1          |1                   |0.50          |0.60        |0.60                 |0.70                
(1 row)

-- multiple resource groups can't share the same name
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
ERROR:  resource group "rg_test_group" already exists

-- cpu_rate_limit/memory_limit range is (0.01, 1)
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.5, memory_limit=.05);
ERROR:  total cpu_rate_limit exceeded the limit of 1.0
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.05, memory_limit=.5);
ERROR:  total memory_limit exceeded the limit of 1.0
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.01, memory_limit=.05);
ERROR:  cpu_rate_limit range is (.01, 1)
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=.05, memory_limit=.01);
ERROR:  memory_limit range is (.01, 1)

-- can't specify the resource limit type multiple times
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, concurrency=1);
ERROR:  Find duplicate resoure group resource type: concurrency
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, cpu_rate_limit=.05);
ERROR:  Find duplicate resoure group resource type: cpu_rate_limit
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, memory_limit=.05);
ERROR:  Find duplicate resoure group resource type: memory_limit
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7, memory_redzone_limit=.8);
ERROR:  Find duplicate resoure group resource type: memory_redzone_limit

DROP RESOURCE GROUP rg_test_group;
DROP

-- can't drop reserved resource groups
DROP RESOURCE GROUP default_group;
ERROR:  cannot drop default resource group "default_group"
DROP RESOURCE GROUP admin_group;
ERROR:  cannot drop default resource group "admin_group"
DROP RESOURCE GROUP none;
ERROR:  resource group "none" does not exist

-- can't drop non-exist resource group
DROP RESOURCE GROUP rg_non_exist_group;
ERROR:  resource group "rg_non_exist_group" does not exist

-- ----------------------------------------------------------------------
-- Test: alter a resource group
-- ----------------------------------------------------------------------
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
CREATE

-- ALTER RESOURCE GROUP SET CONCURRENCY N
-- negative
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -2;
ERROR:  concurrency limit cannot be less than -1
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -0.5;
ERROR:  syntax error at or near "0.5"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -0.5;
                                                            ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0.5;
ERROR:  syntax error at or near "0.5"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0.5;
                                                           ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY a;
ERROR:  syntax error at or near "a"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY a;
                                                           ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 'abc';
ERROR:  syntax error at or near "'abc'"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 'abc';
                                                           ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY '1';
ERROR:  syntax error at or near "'1'"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY '1';
                                                           ^
-- positive
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -1;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 1;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 2;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 1000;
ALTER

DROP RESOURCE GROUP rg_test_group;
DROP

-- ----------------------------------------------------------------------
-- Test: create/alter/drop a resource group in subtransaction
-- ----------------------------------------------------------------------

-- CREATE RESOURCE GROUP cannot run inside a subtransaction
BEGIN;
BEGIN
SAVEPOINT rg_savepoint;
SAVEPOINT
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
ERROR:  CREATE RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK TO SAVEPOINT rg_savepoint;
ROLLBACK
ABORT;
ABORT

-- ALTER RESOURCE GROUP cannot run inside a subtransaction
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
CREATE
SAVEPOINT rg_savepoint;
SAVEPOINT
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 10;
ERROR:  ALTER RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK TO SAVEPOINT rg_savepoint;
ROLLBACK
ABORT;
ABORT

-- DROP RESOURCE GROUP cannot run inside a subtransaction
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=.05, memory_limit=.05, memory_redzone_limit=.7);
CREATE
SAVEPOINT rg_savepoint;
SAVEPOINT
DROP RESOURCE GROUP rg_test_group;
ERROR:  DROP RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK TO SAVEPOINT rg_savepoint;
ROLLBACK
ABORT;
ABORT
